### PR TITLE
Revamp build infrastructure for Clang/LLVM

### PR DESCRIPTION
**Pull request by tony**
Improve finding llvm and clang across platforms. With this patch, the a majority of systems will be able to use their clang/llvm from their package manager without needing to download clang+llvm. This makes it much smoother to install color_coded, as well as much more manageable to handle via ``NeoBundle`` across platforms.

``-DDOWNLOAD_CLANG=1`` can be used to force download.

``-DCUSTOM_CLANG`` is deprecated.

The prior version of clang/link.cmake wasn't reliable in finding LLVM/Clang automatically. It often required entering additional arguments for LLVM/Clang, the locations of which are not consistent across systems, package managers and versions.

FindLibClang can be used to find clang libraries across systems in unison with FindLLVM.

Both support ``${LLVM_ROOT_DIR}``, so the modules will load the libraries in download form.

Tested with both download and packaged libraries (llvm/clang 3.7) on FreeBSD 11-CURRENT,   Ubuntu 15.10 and OS X 10.11 El Capitan.

To test downloading:

``rm -rf ~/.vim/bundle/color_coded/build/*; cmake .. -DDOWNLOAD_CLANG=1 -GNinja && ninja install``

To test with clang/llvm installed from package managers:

``rm -rf ~/.vim/bundle/color_coded/build/*; cmake .. -GNinja && ninja install``

### CODE REVIEW

**Comment by tony** [comment]
I'm not fully sure why tests are failing, other than I can confirm it builds with both clang/llvm from package manager and via download on FreeBSD 11-CURRENT, Ubuntu 15.10 and OS X 10.11.

**Comment by tony** [comment]
I tested this with the instructions on the readme on 12.04:

```
tony@vbox:~/color_coded$ git status
# On branch llvm-cmake-revamp
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
#
#       color_coded_config_test
nothing added to commit but untracked files present (use "git add" to track)
tony@vbox:~/color_coded$ sudo lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 12.04.5 LTS
Release:        12.04
Codename:       precise
tony@vbox:~/color_coded$ make color_coded_test                                                                                                  
[ 88%] Built target color_coded_boost
[100%] Built target color_coded_config_test
running group 'loading and reading of config files'
  test 0 success
  test 1 success
  test 2 success
  test 3 success
  test 4 success
finished group 'loading and reading of config files'
all 5 tests passed
[100%] Built target color_coded_test
```

@jeaye: Any idea what the travis problem could be?

**Comment by tony** [comment]
On Ubuntu 12.04, this worked with both the download and via having clang/llvm installed via package manager.

**Comment by jeaye** [comment]
@tony So, looking at the error on travis:

```bash
$ cd $SOURCEPATH/bin && echo "os.exit()" | lua -l color_coded
lua: error loading module 'color_coded' from file './color_coded.so':
	libclang.so: cannot open shared object file: No such file or directory
stack traceback:
	[C]: ?
	[C]: ?
	[C]: ?
```

This looks like a linking error: color_coded should normally statically link to libclang.a from the downloaded tarball. One issue with the 3.7 release, which is the only reason we didn't upgrade, was that they forgot to include libclang.a... so, it could be that you're using it on that machine and therefore you don't have access to the static lib?

**Comment by jeaye** [comment]
On second thought, @tony, we should move away from static linking. It's easier, on color_coded's front, but it causes issues like #66. So, let's assume we're not going to worry about libclang.a; now we need to worry about libclang.so not being found on travis. Give me a bit to look into this and try it on my machine.

In general, I want to say, **this PR scares me.** Of all the crazy platforms people use, and their different versions and arches, relying on the system to have a proper setup by default seems, to me, the wrong approach. I'm ok with improving the support for using the system's clang, but I'll be very hesitant to have everyone use it. @UnrealQuester, have any input?

**Comment by tony** [comment]
Looks like an rpath problem. It'll work and link correctly before `make install` (`ldd` and `readelf -d` look fine) but after its moved to `bin`, it won't see any the libraries.

**Comment by tony** [comment]
First, whatever the solution is, whether this PR is accepted or not, let me explain the things from a user standpoint.

This is a vim extension, I'd like it to be able to work with ``cmake`` and use the local clang/llvm headers and libs without a hitch. Most distributions package headers for clang and llvm. I see downloading clang+llvm when it already exists as needless duplication and error prone (which I can confirm, since I'm running OS X, FreeBSD, Linux and reinstall it a lot).

> Of all the crazy platforms people use, and their different versions and arches, relying on the system to have a proper setup by default seems, to me, the wrong approach.

It already falls back to downloading if LLVM for Clang isn't found. We can also fallback is versions don't meet minimum requirements.

**Comment by UnrealQuester** [comment]
I would love to have an option to use the system clang libs, but I do not think it should be the default. Many tools that rely on libclang use this approach. This is much less error prone than relying on the system libclang. Some distributions add custom patches and have different versions of clang installed. This makes errors harder to debug for us.

@tony: You mention that the process of downloading clang is error prone. What kind of errors did you encounter?

Also, we might want to try something like [kitchen](https://github.com/test-kitchen/test-kitchen) to test the build on multiple platforms.

**Comment by tony** [comment]
> but I do not think it should be the default. 

I am willing to consider that. But I feel its gratuitous to download clang/llvm on systems where clang/llvm is on the system and in all probability would work.

There will be common platforms we'll be able to detect clang/llvm on and just roll with it.  If we add constraints to the minimum / maximum version of clang, and fallback to downloading for systems known to experience problems, for instance, that could make users' lives easier.

> @tony: You mention that the process of downloading clang is error prone. What kind of errors did you encounter?

Error prone isn't the best way to put it. Downloading clang assures consistency in this respect and it much more deterministic. color_coded is awesome, but entailing downloading clang remotely is a bit much for a vim plugin (if it can be avoided in common cases).

Neobundle times out downloading it. Bump the timeout up, color_coded, out of all the bundles I have installed, takes longer than any other plugin to install. My workaround for that is to use the system's llvm and clang, but since I use OS X, FreeBSD and deb/ubuntu and have color_coded on a lot of systems, it becomes unwieldy to have to pipe in CMake variables to each system.

**Comment by jeaye** [comment]
There's some work going on in YCM right now that may result in color_coded not even having clang as a dependency: I've been pushing to have YCM do the compilation and provide the tokens to color_coded for highlighting. This would simplify so much. Maybe y'all can hop over to https://github.com/Valloric/ycmd/pull/291 and share a similar sentiment.

**Comment by tony** [comment]
YCM is GPLv3. Maintainer [closes threads bringing up licensing / converting to something permissive](https://github.com/Valloric/ycmd/issues/139).

**Comment by jeaye** [comment]
@tony There's now a `3.8.0` branch which upgrades our clang version and switches us over to use clang's shared object. I ran into the same rpath issues with `make install` and fixed it by just not installing to `bin`, but instead the cmake list path. I am interested in cleaning up the LLVM cmake code and better supporting the usage of system clang, but downloading clang has absolutely got to be the default.

Are you interested in updating this with the 3.8.0 branch and making any last changes so I can merge it?

**Comment by tony** [comment]
Rebased against 3.8.0

I have the old version at https://github.com/tony/color_coded/tree/llvm-cmake-revamp-backup just in case.

**Comment by jeaye** [comment]
Looks like the travis changes didn't make it in, since you're still changing into the `bin` directory for the `lua -l color_coded` check, which fails the build.

**Comment by jeaye** [comment]
That change is done here 2ebc9ea6d35b6c605ecdba8f7037a27b98fa2629

File: .travis.yml [hunk 1]
```diff
@@ -34,9 +34,9 @@ install:
...
   - export SOURCEPATH=$PWD
-  - mkdir -p $OUTSOURCE && cd $OUTSOURCE && cmake $SOURCEPATH && make && make install
-  - cd $SOURCEPATH/bin && echo "os.exit()" | lua -l color_coded
-  - cd $SOURCEPATH && rm -rf bin && mv $OUTSOURCE/clang* . && rm -rf $OUTSOURCE
-  - cmake . && make && make install
-  - cd $SOURCEPATH/bin && echo "os.exit()" | lua -l color_coded
+  - mkdir -p $OUTSOURCE && cd $OUTSOURCE && cmake $SOURCEPATH -DDOWNLOAD_CLANG=1 -DMAKE_INSTALL_RPATH_USE_LINK_PATH=ON && make && make install
+  - cd $SOURCEPATH && echo "os.exit()" | lua -l color_coded
+  - cd $SOURCEPATH && mv $OUTSOURCE/clang* . && rm -rf $OUTSOURCE
+  - cmake . -DDOWNLOAD_CLANG=1 -DMAKE_INSTALL_RPATH_USE_LINK_PATH=ON && make && make install
```
Comment (jeaye) [inline .travis.yml]: We set `CMAKE_INSTALL_RPATH_USE_LINK_PATH` always now, so this bit can be removed, I think.

File: CMakeLists.txt [hunk 2]
```diff
@@ -48,21 +48,38 @@ find_package(ZLIB REQUIRED)
 find_package(Threads REQUIRED)
 
 # Clang
-include(cmake/clang/download.cmake)
+if (NOT DOWNLOAD_CLANG)
```
Comment (jeaye) [inline CMakeLists.txt]: We need to default `DOWNLOAD_CLANG` to `TRUE`.

File: CMakeLists.txt [hunk 3]
```diff
@@ -48,21 +48,38 @@ find_package(ZLIB REQUIRED)
...
+	include(cmake/clang/download.cmake)
+	find_package(LLVM)
+	find_package(LibClang)
+endif()
 
 include_directories(
   include
+  ${LLVM_INCLUDE_DIRS}
+  include
+  ${CLANG_INCLUDE_DIRS}
```
Comment (jeaye) [inline CMakeLists.txt]: `include` is here twice, for some reason.

**Comment by tony** [comment]
@jeaye better?

**Comment by jeaye** [comment]
The `-DMAKE_INSTALL_RPATH_USE_LINK_PATH=ON` should be able to be removed completely, since we set it unconditionally now.

File: CMakeLists.txt [hunk 4]
```diff
@@ -48,21 +50,40 @@ find_package(ZLIB REQUIRED)
 find_package(Threads REQUIRED)
 
 # Clang
-include(cmake/clang/download.cmake)
+if (${DOWNLOAD_CLANG})
+  set(DOWNLOAD_CLANG {$DOWNLOAD_CLANG})
+endif()
```
Comment (jeaye) [inline CMakeLists.txt]: I don't understand the meaning of these three lines. What's the point in doing this?

**Comment by tony** [comment]
Sorry for delay, back on it.

**Comment by tony** [comment]
Better?

**Comment by jeaye** [comment]
@tony Will you explain to me why we don't always leave the rpath link option set to true? The way it seems, `DOWNLOAD_CLANG` defaults to true, but you removed the bit that keeps `CMAKE_INSTALL_RPATH_USE_LINK_PATH` true. By default, if we're downloading clang, we need the rpath link path; I'm not sure if we want it when using the system clan though.

File: CMakeLists.txt (hunk 2 above)
Comment (tony) [inline CMakeLists.txt]: If clang LLVM (and the versions we want to support) aren't found - the script defaults to downloading it.

You should you want it to download by default?

**Comment by tony** [comment]
For those who choose to use LLVM/Clang from their systems, what is the minimum version of LLVM/Clang color_coded will be supporting?  Is it 3.8 now?

**Comment by jeaye** [comment]
@tony Technically, 3.6 should still work, but I would certainly recommend anyone upgrade to 3.8.

**Comment by jeaye** [comment]
This is looking great now, @tony, thanks for handling my nitpicks. There's one last thing that would be awesome, and you can let me know if you're willing to do it: the travis build was changed to specify `DOWNLOAD_CLANG=1` which is not actually needed anymore, since it's on by default. Aside from that, it would be good to actually test the usage of system clang in the travis build.

Let me now if you're up for it. Thanks again!

**Comment by tony** [comment]
@jeaye

Revamped a lot. This should be extendable to handle other clang/llvm versions outside 3.8, but I set it at that for now.

The matrix tests for System LLVM 3.8 as well as downloading it, with both GCC and Clang as the compiler. It tests it on both OS X and Linux.

The messages in the CMake macros have been updated to clue in better to what llvm/clang libraries have been found.

**Comment by jeaye** [comment]
@tony This looks awesome; as I've said before: you're the man.

Now that I've merged it, it's odd to see that, if I do `DOWNLOAD_CLANG=0 cmake .` it still downloads clang. I can see it working properly on travis; what might I be missing?

**Comment by tony** [comment]
I've had similar problems, intermittently.

@jeaye have you cleared out the build directory first?

What env are you on?

**Comment by jeaye** [comment]
I'm on OSX. I can:

```bash
$ git clean -fd
$ rm -rf CMakeCache.txt CMakeFiles
$ DOWNLOAD_CLANG=0 cmake .
```

and it still tries to download.

**Comment by tony** [comment]
Its a cmake variable, it needs ``-DDOWNLOAD_CLANG``.

I believe you may be mistaking the way I used an environmental variable in travis.yml? That's just for the test matrix. Its still piped in like ``-DDOWNLOAD_CLANG=${DOWNLOAD_CLANG}``.

Can you try ``cmake . -DDOWNLOAD_CLANG=0``?

**Comment by jeaye** [comment]
Ah, you're right; I had in mind what you were doing in the travis build. Alright, I'm downloading llvm here to go through the motions and then I'll add some updated docs. Thanks again.

**Comment by alientechnology** [comment]
So, how is there a way no, to link with system libclang, without downloading? What is the command? cmake . -DDOWNLOAD_CLANG=0 doesn't work, it's still tries to download.

**Comment by jeaye** [comment]
@alientechnology It does work, but you likely need to clear your cmake cache. Either re-clone color_coded into a fresh directory or run `git clean -fd ; rm -rf CMakeCache.txt CMakeFiles` and try again. I'll get documentation for this up tonight.

**Comment by alientechnology** [comment]
I tried. I still tries to download, but i think i found where the problem is, cmake output:

-- Found LLVM: /usr/lib/llvm-3.8 (found version "3.8.1") 
-- Could not find any Clang libraries in /usr/lib/llvm-3.8/lib
and then starts download. In ubuntu clang if you install from clang repo, clang libraries will be in `/usr/lib/clang/3.8/lib/` (symlink to `/usr/lib/llvm-3.8/lib/clang/3.8.1/lib/`), so is there cmake variable, to specify correct path?

**Comment by jeaye** [comment]
You can try specifying LLVM_CONFIG; check out the travis build in this PR for how it's used. @tony will hopefully have more information and I can put it all into docs when we get it sorted.

**Comment by alientechnology** [comment]
Thanks, now it's working.

**Comment by oblitum** [comment]
cc #120, #126.

**Pull request merged**